worker-document/src/main/java/com/hpe/caf/worker/document/tasks/AbstractTask.java

```diff
@@ -104,4 +142,104 @@ public final WorkerResponse handleGeneralFailure(final Throwable failure)
 
     @Nonnull
     protected abstract WorkerResponse handleGeneralFailureImpl(final Throwable failure);
+
+    /**
+     * Load the customization scripts and if there is an exception then try to determine if it is a transient issue.
+     *
+     * @throws DocumentWorkerTransientException if one of the scripts could not be loaded due to a transient issue
+     * @throws InterruptedException if any thread has interrupted the current thread
+     */
+    public void loadScripts() throws DocumentWorkerTransientException, InterruptedException
+    {
+        try {
+            scripts.loadAll();
+        } catch (final ScriptException ex) {
+            final Throwable cause = ex.getCause();
+
+            if (cause instanceof SocketException) {
+                throw new DocumentWorkerTransientException(ex);
+            } else if (cause instanceof DocumentWorkerTransientException) {
+                throw (DocumentWorkerTransientException) cause;
+            } else if (cause instanceof InterruptedException) {
+                throw (InterruptedException) cause;
+            } else {
+                throw new RuntimeException(ex);
+            }
+        }
+    }
+
+    public void raiseProcessTaskEvent() throws DocumentWorkerTransientException, InterruptedException
+    {
+        raiseScriptEvents("onProcessTask", false, new TaskEventObject(this));
+    }
+
+    public boolean raiseBeforeProcessDocumentEvent(final Document document) throws DocumentWorkerTransientException, InterruptedException
+    {
+        final CancelableDocumentEventObject documentEventObj = new CancelableDocumentEventObject(document);
+
+        raiseScriptEventsUntil("onBeforeProcessDocument", false, () -> documentEventObj.cancel, documentEventObj);
+
+        return documentEventObj.cancel;
+    }
+
+    public void raiseProcessDocumentEvent(final Document document) throws DocumentWorkerTransientException, InterruptedException
+    {
+        raiseScriptEvents("onProcessDocument", false, new DocumentEventObject(document));
+    }
+
+    public void raiseAfterProcessDocumentEvent(final Document document) throws DocumentWorkerTransientException, InterruptedException
+    {
+        raiseScriptEvents("onAfterProcessDocument", true, new DocumentEventObject(document));
```

Comment:
Why is the order of events raised after process document reversed?

Reply:
To produce an onion like effect - script 0 is the outer script - its before events run first but its after events run last.  Overall this should make it easier to reason about when there are multiple scripts, as there is much less interleaving.